# 8.2.1 Extract the Wikipedia Movies JSON

## Find the File

In [1]:
import json
import pandas as pd
import numpy as np

In [2]:
file_dir = '/Users/andreadarrah/Desktop/Movies'

## Load the JSON into a List of Dictionaries

To load the raw JSON into a list of dictionaries, we will use the load() method.

REWIND
Remember, when opening files in Python, we want to use the with statement to handle 
the file resource.

Using the with statement, open the Wikipedia JSON file to be read into the variable file,
and use json.load() to save the data to a new variable.

In [3]:
with open(f'{file_dir}/wikipedia-movies.json', mode = 'r') as file:
    wiki_movies_raw = json.load(file)

Here, wiki_movies_raw is now a list of dicts. Before we take a look at the data, 
we should check how many records were pulled in. We can use the len() function 
(see below), which returns 7,311 records.


In [4]:
len(wiki_movies_raw)

7311

Also, we should always take a look at a few individual records just to make sure that the data didn't come in horribly garbled. With a DataFrame, we'd do this with the head() and tail()methods, but with a list of dicts, we need to inspect the records directly.

REWIND
Remember, since we're working with a list, we'll use index slices to select specific chunks of wiki_movies_raw to inspect directly. This is also a great use case for negative index slices.

To see the first five records, use the following:

In [5]:
# First 5 records
wiki_movies_raw[:5]

[{'url': 'https://en.wikipedia.org/wiki/The_Adventures_of_Ford_Fairlane',
  'year': 1990,
  'imdb_link': 'https://www.imdb.com/title/tt0098987/',
  'title': 'The Adventures of Ford Fairlane',
  'Directed by': 'Renny Harlin',
  'Produced by': ['Steve Perry', 'Joel Silver'],
  'Screenplay by': ['David Arnott', 'James Cappe', 'Daniel Waters'],
  'Story by': ['David Arnott', 'James Cappe'],
  'Based on': ['Characters', 'by Rex Weiner'],
  'Starring': ['Andrew Dice Clay',
   'Wayne Newton',
   'Priscilla Presley',
   'Lauren Holly',
   'Morris Day',
   'Robert Englund',
   "Ed O'Neill"],
  'Narrated by': 'Andrew "Dice" Clay',
  'Music by': ['Cliff Eidelman', 'Yello'],
  'Cinematography': 'Oliver Wood',
  'Edited by': 'Michael Tronick',
  'Productioncompany ': 'Silver Pictures',
  'Distributed by': '20th Century Fox',
  'Release date': ['July 11, 1990', '(', '1990-07-11', ')'],
  'Running time': '102 minutes',
  'Country': 'United States',
  'Language': 'English',
  'Budget': '$20 million',


To see the last five records, use the following:

In [6]:
# Last 5 records
wiki_movies_raw[-5:]

[{'url': 'https://en.wikipedia.org/wiki/Holmes_%26_Watson',
  'year': 2018,
  'imdb_link': 'https://www.imdb.com/title/tt1255919/',
  'title': 'Holmes & Watson',
  'Directed by': 'Etan Cohen',
  'Produced by': ['Will Ferrell',
   'Adam McKay',
   'Jimmy Miller',
   'Clayton Townsend'],
  'Screenplay by': 'Etan Cohen',
  'Based on': ['Sherlock Holmes',
   'and',
   'Dr. Watson',
   'by',
   'Sir Arthur Conan Doyle'],
  'Starring': ['Will Ferrell',
   'John C. Reilly',
   'Rebecca Hall',
   'Rob Brydon',
   'Steve Coogan',
   'Ralph Fiennes'],
  'Music by': 'Mark Mothersbaugh',
  'Cinematography': 'Oliver Wood',
  'Edited by': 'Dean Zimmerman',
  'Productioncompanies ': ['Columbia Pictures',
   'Gary Sanchez Productions',
   'Mosaic Media Group',
   'Mimran Schur Pictures'],
  'Distributed by': 'Sony Pictures Releasing',
  'Release date': ['December 25, 2018',
   '(',
   '2018-12-25',
   ')',
   '(United States)'],
  'Running time': '90 minutes',
  'Country': 'United States',
  'Language

It's always a good idea to check records in the middle as well. Choose a number 
somewhere around the halfway mark and look at a handful of records after that index.

In [7]:
# Some records in the middle
wiki_movies_raw[3600:3605]

[{'url': 'https://en.wikipedia.org/wiki/Benji:_Off_the_Leash!',
  'year': 2004,
  'imdb_link': 'https://www.imdb.com/title/tt0315273/',
  'title': 'Benji: Off the Leash!',
  'Directed by': 'Joe Camp',
  'Written by': 'Joe Camp',
  'Starring': ['Benji', 'Nick Whitaker', 'Shaggy', 'Gypsy the Cockatoo'],
  'Music by': 'Antonio di Lorenzo',
  'Productioncompany ': 'Mulberry Square Productions',
  'Distributed by': 'Mulberry Square Productions',
  'Release date': ['March 26, 2004', '(', '2004-03-26', ')'],
  'Running time': '97 min',
  'Country': 'United States',
  'Language': 'English',
  'Box office': '$3,817,362'},
 {'url': 'https://en.wikipedia.org/wiki/The_Best_Thief_in_the_World',
  'year': 2004,
  'imdb_link': 'https://www.imdb.com/title/tt0389796/',
  'title': 'The Best Thief in the World',
  'Directed by': 'Jacob Kornbluth',
  'Produced by': ['Tim Perrell', 'Nicola Usborne'],
  'Written by': 'Jacob Kornbluth',
  'Starring': ['Marc Rozendaal',
   'Michael Silverman',
   'David Warsh

# 8.2.2 Extract the Kaggle Data

Since the Kaggle data is already in flat-file formats, we'll just pull them into Pandas DataFrames directly with the following code.

In [8]:
movies_metadata = pd.read_csv('../../Movies/movies_metadata.csv', low_memory=False)
ratings = pd.read_csv('../../Movies/ratings.csv')

Inspect the two DataFrames using the head(), tail(), and sample() methods to make sure that everything seems to be loaded in correctly. (We'll do a deeper dive in the Transform step.)

In [9]:
movies_metadata.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [10]:
movies_metadata.tail()

adult belongs_to_collection budget  \
45461  False                   NaN      0   
45462  False                   NaN      0   
45463  False                   NaN      0   
45464  False                   NaN      0   
45465  False                   NaN      0   

                                                  genres  \
45461  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...   
45462                      [{'id': 18, 'name': 'Drama'}]   
45463  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   
45464                                                 []   
45465                                                 []   

                                   homepage      id    imdb_id  \
45461  http://www.imdb.com/title/tt6209470/  439050  tt6209470   
45462                                   NaN  111109  tt2028550   
45463                                   NaN   67758  tt0303758   
45464                                   NaN  227506  tt0008536   
45465                                   NaN  461257  tt6980792   

      original_language       original_title  \
45461                fa              رگ خواب   
45462                tl  Siglo ng Pagluluwal   
45463                en             Betrayal   
45464                en  Satana likuyushchiy   
45465                en             Queerama   

                                                overview  ... release_date  \
45461        Rising and falling between a man and woman.  ...          NaN   
45462  An artist struggles to finish his work while a...  ...   2011-11-17   
45463  When one of her hits goes wrong, a professiona...  ...   2003-08-01   
45464  In a small town live two brothers, one a minis...  ...   1917-10-21   
45465  50 years after decriminalisation of homosexual...  ...   2017-06-09   

      revenue runtime                          spoken_languages    status  \
45461     0.0    90.0    [{'iso_639_1': 'fa', 'name': 'فارسی'}]  Released   
45462     0.0   360.0         [{'iso_639_1': 'tl', 'name': ''}]  Released   
45463     0.0    90.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   
45464     0.0    87.0                                        []  Released   
45465     0.0    75.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                          tagline                title  video  \
45461  Rising and falling between a man and woman               Subdue  False   
45462                                         NaN  Century of Birthing  False   
45463                      A deadly game of wits.             Betrayal  False   
45464                                         NaN     Satan Triumphant  False   
45465                                         NaN             Queerama  False   

      vote_average vote_count  
45461          4.0        1.0  
45462          9.0        3.0  
45463          3.8        6.0  
45464          0.0        0.0  
45465          0.0        0.0  

[5 rows x 24 columns]

In [11]:
movies_metadata.sample(n=5)

adult belongs_to_collection budget  \
15204  False                   NaN      0   
1900   False                   NaN      0   
38965  False                   NaN      0   
8987   False                   NaN      0   
21582  False                   NaN      0   

                                                  genres homepage     id  \
15204                                                 []      NaN  21245   
1900   [{'id': 9648, 'name': 'Mystery'}, {'id': 878, ...      NaN  12101   
38965  [{'id': 9648, 'name': 'Mystery'}, {'id': 16, '...      NaN  14233   
8987                                                  []      NaN  53688   
21582  [{'id': 9648, 'name': 'Mystery'}, {'id': 878, ...      NaN  86235   

         imdb_id original_language                    original_title  \
15204  tt0307076                fi                        Kuutamolla   
1900   tt0070723                en                     Soylent Green   
38965  tt0970933                en                         Film Noir   
8987   tt0102982                sv  Stinsen brinner... filmen alltså   
21582  tt1031276                en                   Bad Girl Island   

                                                overview  ... release_date  \
15204                                                NaN  ...   2002-02-01   
1900   In an overpopulated futuristic Earth, a New Yo...  ...   1973-04-19   
38965  Private detective Sam Ruben's clever plan fall...  ...   2007-03-02   
8987                                                 NaN  ...   1991-03-03   
21582  A wealthy film producer Michael Pace discovers...  ...   2009-01-01   

         revenue runtime                          spoken_languages    status  \
15204        0.0   117.0    [{'iso_639_1': 'fi', 'name': 'suomi'}]  Released   
1900   3600000.0    97.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   
38965        0.0    97.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   
8987         0.0     0.0  [{'iso_639_1': 'sv', 'name': 'svenska'}]  Released   
21582        0.0    90.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                    tagline                             title  \
15204                                   NaN                  Lovers & Leavers   
1900   What is the secret of Soylent Green?                     Soylent Green   
38965                                   NaN                         Film Noir   
8987                                    NaN  Stinsen brinner... filmen alltså   
21582   Stay out of the water... She bites.                   Bad Girl Island   

       video vote_average vote_count  
15204  False          6.0        4.0  
1900   False          6.8      305.0  
38965  False          6.3        5.0  
8987   False          5.8        2.0  
21582  False          4.7        6.0  

[5 rows x 24 columns]

In [12]:
ratings.head()

userId  movieId  rating   timestamp
0       1      110     1.0  1425941529
1       1      147     4.5  1425942435
2       1      858     5.0  1425941523
3       1     1221     5.0  1425941546
4       1     1246     5.0  1425941556

In [13]:
ratings.tail()

userId  movieId  rating   timestamp
26024284  270896    58559     5.0  1257031564
26024285  270896    60069     5.0  1257032032
26024286  270896    63082     4.5  1257031764
26024287  270896    64957     4.5  1257033990
26024288  270896    71878     2.0  1257031858

In [14]:
ratings.sample(n=5)

userId  movieId  rating   timestamp
20695745  215058     2836     4.0  1011670976
25811708  268536     3062     4.0   944924159
3418742    35552      608     4.0   833505023
22405876  232517     7930     2.0  1162545546
4723350    48543    68157     4.0  1442087959

# 8.3.1 Data-Cleaning Strategies

Wikipedia doesn't have strict standards on how movie data is presented, so it needs 
a lot of work to clean up the data and make it usable. Like most web-scraped data, 
it's in the flexible JSON format to store all kinds of data, but Britta needs to 
organize it in a structured format before she can send it to SQL—and she's asked you 
to assist with this task. (You do have experience in this, after all.) First, explore 
your options for cleaning the dataset.

The transform step is largely spent on data cleaning. There are other transformations
that aren't strictly data cleaning, but for the most part, the transformation step is
used to clean up your data.

Every messy dataset presents its own unique challenges. There's no one right way to clean data, but we can still have a rough game plan to follow.

Bad data comes in three states:

Beyond repair
Badly damaged
Wrong form
The state of the data largely determines which strategy you should use to clean it.

Data beyond repair could be data that has been overwritten or has suffered severe data corruption during storage or transfer (such as power loss during writing, voltage spikes, or hard-drive failures). The worst-case example would be having data with every value missing. All the information is lost and unrecoverable. For data beyond repair, all we can do is delete it and move on.

Data that is badly damaged may have good data that we can recover, but it will take time and effort to repair the damaged data. This can be garbled data, with a lot of missing values, from inconsistent sources, or existing in multiple columns. Consider trade-offs to pick the best solution (even if the "best" solution isn't perfect, but rather the "best-available" solution). To repair badly damaged data, try these strategies:

Filling in missing data by
substituting data from another source,
interpolating between existing data points, or
extrapolating from existing data
Standardizing units of measure (e.g., monetary values stored in multiple currencies)
Consolidating data from multiple columns
Finally, data in the wrong form should usually be fixed—that is, the data is good but can't be used in its current form. "Good" data in the wrong form can be data that is too granular or detailed, numeric data stored as strings, or data that needs to be split into multiple columns (e.g., address data). To remedy good data in the wrong form, try these strategies:

Reshape the data
Convert data types
Parse text data to the correct format
Split columns
These options are all available to us, but knowing when to perform which strategy can feel overwhelming. There is no simple checklist or flowchart we can use to guide us, and ultimately, that's a good thing. In data cleaning, we have to constantly ask ourselves what we might have missed, and following a rigid plan means we won't be asking ourselves those important questions. Data cleaning requires a lot of improvising.

IMPORTANT
It's important to document your data cleaning assumptions as well as decisions and their motivations. Later decisions depend on earlier decisions made, which can be too much to remember. Any assumptions that were part of an earlier decision can, if forgotten, ruin later steps.

Transforming a messy dataset into a clean dataset is an iterative process. As you clean one part of the data, you may reveal something messy in another part of the data. Sometimes that means unwinding a lot of work that you've already done and having to redo it with a slight change. Documenting why a particular step is necessary will show you how to redo it without introducing more errors.

We're not completely lost—we do have a strategy. We're not going to try and clean all the data at once. Instead, we're going to focus on one problem at a time using an iterative process.

# 8.3.2 Iterative Process for Cleaning Data

The iterative process for cleaning data can be broken down as follows:

First, we need to inspect our data and identify a problem.
Once we've identified the problem, we need to make a plan and decide whether it is worth the time and effort to fix it.
Finally, we execute the repair.
Early iterations focus on making the data easier to investigate: deleting obviously bad data, removing superfluous columns (e.g., columns with only one value or missing an overwhelming amount of data), removing duplicate rows, consolidating columns, and reshaping the data if necessary.

As the data becomes easier to investigate, iterations focus on fixing the most obvious problems first. As obvious problems are resolved, more subtle problems become noticeable.

As the iterations shift toward solving more subtle problems, we might discover an earlier step needs to change as well as all the iterations that follow that step. It's frustrating when work has to be undone, but at least you now have a better understanding of your data.

NOTE
In general, earlier iterations try to handle big chunks of data at one time, such as removing columns and rows, while later iterations focus on smaller chunks of data, such as parsing values.

It's rare to reach a point where no more problems exist in the data. More likely, a point is reached where the work to fix any remaining problems isn't worth the amount of data that would be recovered. After the remaining issues are documented, the transform step is considered finished.

Now that we know how to use our iterative process, let's review each step in detail.

#### Inspect portion of the process

Before we can do anything, we have to look at our data. The first thing we want to know is whether or not the data was imported correctly. The simplest way to confirm this is to print out the first few data points and examine the first few rows for irregularities, e.g., data in the wrong columns, all missing values, column headers that don't make sense, or garbled characters.

If the data doesn't look correct, we know it wasn't imported correctly. Sometimes the beginning of the data looks fine, but if the import went wrong somewhere in the middle of the process, the rest of the data can be affected.

Therefore, it's good practice to check the last few rows and a random sample of rows. We can also start to answer some simple questions about the data:

Does it have a consistent structure (like a CSV table) or is it unstructured (like a collection of email messages)?

How is each data point identified—is there an explicit, unique ID for each data point, or will one need to be built?

However, most usable data contains too many data points to review every single one, so we'll need to use strategies that tell us about the whole dataset.

First, count how many data points or rows exist. If the data is structured, count the number of columns and missing values in each column. If possible, count the number of unique values in each column and how frequently each unique value appears. To determine if this is possible, we'll need to investigate the data types for each column.

When investigating the data type for a column, we want to know what the data type is and what the data type should be. For example, if we see "True" and "False" as entries for a column, we expect that the data type will be a Boolean. If the data type is a string, we need to investigate further.

If a column's data type is numeric, we can summarize its data with some basic statistics, such as measures of central tendency (e.g., mean and/or median) and measures of spread (e.g., standard deviation, interquartile range, minimum/maximum). We can also investigate columns with statistical plots, like scatter plots and histograms.

#### Plan portion of the process

After we've investigated our data and started to identify problem areas, we can make decisions about how to fix the problems. This requires articulating the problems clearly—even if that is simply expressing the problems to ourselves—and devising a plan to modify the data and fix the problem. In this step, we'll answer several questions, including:

If a column doesn't have the right data type, is it a problem with the whole column? Or are just a handful of rows causing the issues?

Do rows have outliers due to spurious data? Or are they valid data points?

When values are missing, will they need to be removed, replaced, or interpolated?

The answers to these questions will tell us how we need to modify our data. Keep in mind, there are two main ways: we can modify values and we can modify structure.

Modifying data values includes removing rows or columns, replacing values, or generating new columns from old ones. We might remove rows with missing or corrupted data, columns with only one value, or columns mostly missing data. There are many ways we might replace data. Instead of dropping missing values, we might replace them with zeros or empty strings. We might have a column that contains nonstandard values, such as percentages that are stored as whole numbers from 0 to 100 and also as fractions from 0 to 1, and we would replace them with one standard form.

Converting a column to a new data type is also a form of replacing values. We can also bin data (like rounding to the nearest hundred), replacing numeric data (e.g., income) with categorical data (e.g., income brackets). We might generate new columns by splitting an existing column into several new columns—by splitting an address column to street, city, state, and zip code columns, for example—or by calculating a new column from multiple existing columns, like calculating total price by multiplying item prices by quantities.

Modifying data structure includes pivoting the values of one column into multiple columns, aggregating rows, and merging multiple data sets. It can also include aggregating large amounts of data into summary data or summary statistics.

With clearly stated steps to fix the problem, we can make an informed decision about whether implementing the plan is worth the effort. Sometimes there are multiple viable resolutions to choose from. To decide, we weigh trade-offs and ultimately choose the best option.

#### Execute portion of the process

Once we have a detailed list of steps to modify our dataset, it's time to implement it. We'll start writing code to fix the problem we're focusing on.

As we write, we might discover that the problem is more difficult than initially expected. This is a normal part of the process. As you implement your changes, try to take into account any unintended consequences you could introduce.

After implementing your changes, the next step is to return and inspect the data in a new iteration. This step is important, especially when modifying data structure, which can introduce missing data points, or inadvertently create more bad data.

### Cleanup Is Messy Work

While transforming your data, you might bounce between steps in the iteration—for example, making a plan, then realizing you need to inspect more; executing a plan, then realizing a step was missed and you need to quickly rework the plan. We offer these steps as a descriptive, not prescriptive, approach. Cleaning up messy data is a messy process. The best practice is to document every step of your thought process and actions in detail.

Now let's go get our hands dirty with some messy data.

# 8.3.3 Investigating the Wikipedia Data

## Initial Investigation

### Inspect

One of the easiest ways to find glaring errors is to just pretend as if there aren't any, and try to jump straight to the finish line. Eventually, we want to clean up the Wikipedia data into tabular data with rows and columns, so let's see what happens if we create a DataFrame from our raw data.

In [15]:
wiki_movies_df = pd.DataFrame(wiki_movies_raw)

In [16]:
wiki_movies_df.head()

url    year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990.0   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990.0   
2   https://en.wikipedia.org/wiki/Air_America_(film)  1990.0   
3    https://en.wikipedia.org/wiki/Alice_(1990_film)  1990.0   
4      https://en.wikipedia.org/wiki/Almost_an_Angel  1990.0   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2  https://www.imdb.com/title/tt0099005/                      Air America   
3  https://www.imdb.com/title/tt0099012/                            Alice   
4  https://www.imdb.com/title/tt0099018/                  Almost an Angel   

          Directed by                  Produced by  \
0        Renny Harlin   [Steve Perry, Joel Silver]   
1         James Foley  [Ric Kidney, Robert Redlin]   
2  Roger Spottiswoode               Daniel Melnick   
3         Woody Allen              Robert Greenhut   
4        John Cornell                 John Cornell   

                                Screenplay by                     Story by  \
0  [David Arnott, James Cappe, Daniel Waters]  [David Arnott, James Cappe]   
1                [James Foley, Robert Redlin]                          NaN   
2                  [John Eskow, Richard Rush]                          NaN   
3                                         NaN                          NaN   
4                                         NaN                          NaN   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   
2             [Air America, by, Christopher Robbins]   
3                                                NaN   
4                                                NaN   

                                            Starring  ... Predecessor  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...  ...         NaN   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...  ...         NaN   
2  [Mel Gibson, Robert Downey Jr., Nancy Travis, ...  ...         NaN   
3  [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...  ...         NaN   
4        [Paul Hogan, Elias Koteas, Linda Kozlowski]  ...         NaN   

  Founders Area served Products Services Russian Hebrew Revenue  \
0      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
1      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
2      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
3      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
4      NaN         NaN      NaN      NaN     NaN    NaN     NaN   

  Operating income Polish  
0              NaN    NaN  
1              NaN    NaN  
2              NaN    NaN  
3              NaN    NaN  
4              NaN    NaN  

[5 rows x 193 columns]

We usually use head() to inspect a few rows of data, but it also tells us about the shape of our DataFrame. Below the rows of data, you'll see that it says there are 5 rows of data and 193 columns. That's a lot of columns! Even if we try to use print(wiki_movies_df.columns), they won't all print out. We'll have to convert wiki_movies_df.columns to a list to see all of the columns.

Use wiki_movies_df.columns.tolist() and run the cell to see all of the column names that were imported. Your output should appear as follows:

In [17]:
wiki_movies_df.columns.tolist()

['url',
 'year',
 'imdb_link',
 'title',
 'Directed by',
 'Produced by',
 'Screenplay by',
 'Story by',
 'Based on',
 'Starring',
 'Narrated by',
 'Music by',
 'Cinematography',
 'Edited by',
 'Productioncompany ',
 'Distributed by',
 'Release date',
 'Running time',
 'Country',
 'Language',
 'Budget',
 'Box office',
 'Written by',
 'Genre',
 'Theme music composer',
 'Country of origin',
 'Original language(s)',
 'Producer(s)',
 'Editor(s)',
 'Production company(s)',
 'Original network',
 'Original release',
 'Productioncompanies ',
 'Executive producer(s)',
 'Production location(s)',
 'Distributor',
 'Picture format',
 'Audio format',
 'Voices of',
 'Followed by',
 'Composer(s)',
 'Created by',
 'Also known as',
 'Opening theme',
 'No. of episodes',
 'Preceded by',
 'Author',
 'Publisher',
 'Publication date',
 'Media type',
 'Pages',
 'ISBN',
 'OCLC',
 'LC Class',
 'Cover artist',
 'Series',
 'Set in',
 'Adaptation by',
 'Suggested by',
 'Biographical data',
 'Born',
 'Died',
 'Resti

We can identify column names that don't relate to movie data, such as "Dewey Decimal," "Headquarters," and "Number of employees." (There may be other examples that jumped out at you as well.)

### Plan

We can identify column names that don't relate to movie data, such as "Dewey Decimal," "Headquarters," and "Number of employees." (There may be other examples that jumped out at you as well.)

## Use List Comprehentions to Filter Data

#### REWIND

We've used list comprehensions previously as a compact way to apply a function to every element in a list. In this module, we'll use list comprehensions to filter data.

We've used list comprehensions previously as a compact way to apply a function to every element in a list. In this module, we'll use list comprehensions to filter data.

So far, we've used list comprehensions in the form to compress code that would have been done in a for loop.

[expression for element in source_list]

We can also filter out results using a conditional filter expression, as shown below:

[expression for element in source_list if filter_expression]

The resulting list will only have elements where the filter expression evaluates to True.

To create a filter expression for only movies with a director and an IMDb link, keep in mind that there are two columns in the data for director information. We'll need to check if either "Director" or "Directed by" are keys in the current dict. If there is a director listed, we also want to check that the dict has an IMDb link. Luckily, that information is only in one column, imdb_link, so our filter expression will look like the following:

if ('Director' in movie or 'Directed by' in movie) and 'imdb_link' in movie

### Execute

Create a list comprehension with the filter expression we created and save that to an intermediate variable wiki_movies. See how many movies are in wiki_movies with the len() function.

In [18]:
wiki_movies = [movie for movie in wiki_movies_raw
              if ('Director' in movie or 'Directed by' in movie)
              and 'imdb_link' in movie]
len(wiki_movies)

7080

In [19]:
wiki_imdb_director_df = pd.DataFrame(wiki_movies)
wiki_imdb_director_df.head()

url  year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990   
2   https://en.wikipedia.org/wiki/Air_America_(film)  1990   
3    https://en.wikipedia.org/wiki/Alice_(1990_film)  1990   
4      https://en.wikipedia.org/wiki/Almost_an_Angel  1990   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2  https://www.imdb.com/title/tt0099005/                      Air America   
3  https://www.imdb.com/title/tt0099012/                            Alice   
4  https://www.imdb.com/title/tt0099018/                  Almost an Angel   

          Directed by                  Produced by  \
0        Renny Harlin   [Steve Perry, Joel Silver]   
1         James Foley  [Ric Kidney, Robert Redlin]   
2  Roger Spottiswoode               Daniel Melnick   
3         Woody Allen              Robert Greenhut   
4        John Cornell                 John Cornell   

                                Screenplay by                     Story by  \
0  [David Arnott, James Cappe, Daniel Waters]  [David Arnott, James Cappe]   
1                [James Foley, Robert Redlin]                          NaN   
2                  [John Eskow, Richard Rush]                          NaN   
3                                         NaN                          NaN   
4                                         NaN                          NaN   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   
2             [Air America, by, Christopher Robbins]   
3                                                NaN   
4                                                NaN   

                                            Starring  ... Hepburn Literally  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...  ...     NaN       NaN   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...  ...     NaN       NaN   
2  [Mel Gibson, Robert Downey Jr., Nancy Travis, ...  ...     NaN       NaN   
3  [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...  ...     NaN       NaN   
4        [Paul Hogan, Elias Koteas, Linda Kozlowski]  ...     NaN       NaN   

  Cantonese Chinese Yiddish Arabic Romanized Russian Hebrew Polish  
0       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
1       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
2       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
3       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
4       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  

[5 rows x 78 columns]

### Inspect

#### Important
One thing to watch out for is to make nondestructive edits as much as possible while designing your pipeline. That means it's better to keep your raw data in one variable, and put the cleaned data in another variable. It takes up more memory, but it makes tracking the iterative process of data cleaning easier.

For example, if we had just deleted the movies from wiki_movies_raw that didn't have "Directed by" as a key, we'd have made a destructive edit and missed that some have "Director" as the key instead. This can cause errors to creep in until we realize our mistake, and if we made destructive edits, it would be impossible to see what caused those errors. Using nondestructive edits helps determine the origin of errors.

There sure are a lot of languages—we'll get to those shortly. For now, one of the columns that stands out is "No. of episodes."

### Plan

It looks like we've got some TV shows in our data instead of movies. We'll want to get rid of those, too.

### Execute

We'll add that filter to our list comprehension.

In [20]:
wiki_movies = [movie for movie in wiki_movies_raw
              if ('Director' in movie or 'Directed by' in movie)
              and 'imdb_link' in movie
              and 'No. of episodes' not in movie]

#### IMPORTANT
Don't worry if you didn't catch the "No. of episodes" column in the list. Cleaning data is an iterative process, and if you started with cleaning up the language data first, or some other part of the data, you would see the "No. of episodes" column soon enough. The key is to keep reworking the pipeline bit by bit.

# 8.3.4 Revisit Functions

To keep things organized, we are going to make a function that performs that cleaning process.

#### REWIND
Remember, functions are blocks of code within a script or algorithm that perform a specific task. There are four basic parts to a function:

Name
Parameters
Code block
Return value

We're going to expand on some trickier aspects of functions before we make our movie cleaning function.

First, we need to talk about scope. Inside the code block of a function, we can use variables that were created outside the function and initialize new variables inside the function.

This is called the "scope" of the variables:

Variables created outside the function are called global variables.

New variables created inside the function are local variables.

The hierarchy of variables is called the scope.

#### IMPORTANT
The scope of local variables only works inside the function in which they are created. If we try to access a local variable outside the function in which it's defined, Python will raise a NameError because it won't be able to find the function.

Keep this trick in mind with the scope of variables: We can create a local variable with the same name as a global variable. Inside the function, the local variable takes precedence, but the value of the global variable will be unchanged. For example, consider the code below:

x = 'global value'

def foo():
    x = 'local value'
    print(x)

foo()
print(x)

The output of this code would be:

local value
global value

Outside the function, x has the value global value. Although we define a new function after assigning a value to the global variable x, the x inside the function is a new variable whose scope is local to foo(). Even after calling the foo() function, the global variable x keeps its original value.

However, we have to be careful with variables we send to a function. Any data structure can be sent as a parameter to a function, including dicts, lists, and tuples, but we must be careful with mutable objects.

#### CAUTION
When passing mutable objects like a dict or list as parameters to a function, the function can change the values inside the object.

For example, consider the code below:

my_list = [1,2,3]
def append_four(x):
    x.append(4)
append_four(my_list)
print(my_list)

The output would be:

[1, 2, 3, 4]

It changed my_list to [1,2,3,4], even outside the function. To leave the original object unchanged, we need to make a copy. For lists, we make a copy with the list function; for dicts, we make a copy with the dict function, like so:

new_list = list(old_list)
new_dict = dict(old_dict)

### Lambda Functions

There's a special function we can make in Python called lambda, which is the most stripped-down kind we can make.

Lambda functions are written in one line and automatically return a value without using the return keyword. Lambda functions have no name and are also known as "anonymous functions."

So, how do we call a function with no name? We don't. There are functions that expect other functions to be sent to them as a parameter, and lambda functions are a way to quickly create a concise function to send as a parameter to another function. We'll return to lambda functions in a later section, but the basic syntax to a lambda function follows:

lambda arguments: expression

This function will take in an argument and will return the expression. Even though lambda functions are supposed to be anonymous, just this once we'll create a lambda function and assign a name so that we can see how they work. A lambda function that squares a value looks like the following:

lambda x: x * x

Here, x is the argument, and x * x is the expression. Let's assign this to a name so that we can use it:

square = lambda x: x * x
square(5)

The output will be:
25

If this seems strange, don't worry. The main benefit of lambda functions is that they can be used as one-time-use functions. We'll talk more about lambda functions when we have a more natural use case for them—they'll make more sense at that time.

#### NOTE

There are many esoteric topics related to functions, including defining functions within functions, recursion, and functions that create more functions. These are part of the functional programming paradigm. Functional programming has its place, but it also has a well-deserved reputation for being confusing and a bit niche. It's good to be aware of functional programming, but we won't rely on its concepts for this module.

Functions are powerful, but enough talk—let's put them to work.

# 8.3.5 Create a Function to Clean the Data, Part 1

Filtering out bad data isn't enough. You know that you need to make sure the good data that you have is clean enough to use. There's a lot at stake! 
Now we're ready to create our function to clean our movie data.

First, write a simple function to make a copy of the movie and return it. As we work with our data, we'll iteratively add more to our code block. To start, call the function clean_movie, and have it take movie as a parameter.

#### def clean_movie(movie):

Because the movies are dicts and we want to make nondestructive edits, make a copy of the incoming movie.

To make a copy of movie, we'll use the dict() constructor.

#### IMPORTANT
Constructors are special functions that initialize new objects. They reserve space in memory for the object and perform any initializations the object requires. Also, constructors can take parameters and initialize a new object using those parameters.

When we pass movie as a parameter to the dict() constructor, it reserves a new space in memory and copies all of the info in movie to that new space.

As an example, we could start our function off with this code:

#### def clean_movie(movie):
    movie_copy = dict(movie)
    
However, we have another trick that's even better.

Inside of the function, we can create a new local variable called movie and assign it the new copy of the parameter movie.

#### def clean_movie(movie):
    movie = dict(movie) #create a non-destructive copy
This way, inside of the clean_movie() function, movie will refer to the local copy. Any changes we make inside clean_movie() will now only affect the copy, so if we make a mistake, we still have the original, untouched movie to reference.

To finish our skeleton of the clean_movie function, return the movie variable.

In [21]:
def clean_movie(movie):
    movie = dict(movie) #create a non-destructive copy
    return movie

### Inspect 
Now take a look at what's going on with those languages. The first one on the list is Arabic, so let's see which movies have a value for "Arabic."

In [22]:
wiki_movies_df[wiki_movies_df['Arabic'].notnull()]['url']

7060    https://en.wikipedia.org/wiki/The_Insult_(film)
7293     https://en.wikipedia.org/wiki/Capernaum_(film)
Name: url, dtype: object

### Plan
The different language columns are for alternate titles of the movie. Let's combine all of them into one dictionary that has all the alternate titles.

To do that, we need to go through each of the columns, one by one, and determine which are alternate titles. Some might be tricky. If you're not sure what a column name means, google it. Also, review a column's data to understand the type of content in that column.

For example, perhaps you've never heard of "McCune–Reischauer." Is it an esoteric filmmaking technique? Google it, and you'll learn it's a romanization system for Korean. Now look at the actual values contained in the column. If the values don't make sense to you either, google them, too.

#### NOTE
The value_counts() method is a quick, easy way to see what non-null values there are in a column.

Try the following Skill Drill. If you're not sure, don't guess. Look at the data, and investigate the source if you have any questions. There are no shortcuts in this task.

In [23]:
sorted(wiki_movies_df.columns.tolist())

['Actor control',
 'Adaptation by',
 'Alias',
 'Alma mater',
 'Also known as',
 'Animation by',
 'Arabic',
 'Area',
 'Area served',
 'Artist(s)',
 'Attraction type',
 'Audio format',
 'Author',
 'Based on',
 'Biographical data',
 'Bopomofo',
 'Born',
 'Box office',
 'Budget',
 'Camera setup',
 'Cantonese',
 'Characters',
 'Children',
 'Chinese',
 'Cinematography',
 'Closing date',
 'Color process',
 'Comics',
 'Composer(s)',
 'Coordinates',
 'Country',
 'Country of origin',
 'Cover artist',
 'Created by',
 'Date premiered',
 'Designer(s)',
 'Developed by',
 'Developer(s)',
 'Dewey Decimal',
 'Died',
 'Directed by',
 'Director',
 'Distributed by',
 'Distributor',
 'Divisions',
 'Duration',
 'Edited by',
 'Editor(s)',
 'Ending theme',
 'Engine',
 'Engine(s)',
 'Executive producer(s)',
 'Family',
 'Fate',
 'Film(s)',
 'Followed by',
 'Format(s)',
 'Formerly',
 'Founded',
 'Founder',
 'Founders',
 'French',
 'Full name',
 'Gender',
 'Genre',
 'Genre(s)',
 'Genres',
 'Gwoyeu Romatzyh',
 'Ha

## Handle the Alternative Titles

### Plan

Now we can add in code to handle the alternative titles. The logic we need to implement follows:

Make an empty dict to hold all of the alternative titles.
Loop through a list of all alternative title keys:
Check if the current key exists in the movie object.
If so, remove the key-value pair and add to the alternative titles dict.
After looping through every key, add the alternative titles dict to the movie object.

### Execute
#### Step 1: Make an empty dict to hold all of the alternative titles.

In [32]:
def clean_movie(movie):
    movie = dict(movie) #create a non-destructive copy
    alt_titles = {}
    return movie

#### Step 2: Loop through a list of all alternative title keys.

In [35]:
def clean_movie(movie):
    movie = dict(movie) #create a non-destructive copy
    alt_titles = {}
    for key in ['Also known as','Arabic','Cantonese','Chinese','French',
                'Hangul','Hebrew','Hepburn','Japanese','Literally',
                'Mandarin','McCune–Reischauer','Original title','Polish',
                'Revised Romanization','Romanized','Russian',
                'Simplified','Traditional','Yiddish']:

        return movie

#### Step 2a: Check if the current key exists in the movie object.

In [37]:
def clean_movie(movie):
    movie = dict(movie) #create a non-destructive copy
    alt_titles = {}
    for key in ['Also known as','Arabic','Cantonese','Chinese','French',
                'Hangul','Hebrew','Hepburn','Japanese','Literally',
                'Mandarin','McCune–Reischauer','Original title','Polish',
                'Revised Romanization','Romanized','Russian',
                'Simplified','Traditional','Yiddish']:
        if key in movie:

            return moviee

#### Step 2b: If so, remove the key-value pair and add to the alternative titles dictionary.

In [38]:
def clean_movie(movie):
    movie = dict(movie) #create a non-destructive copy
    alt_titles = {}
    for key in ['Also known as','Arabic','Cantonese','Chinese','French',
                'Hangul','Hebrew','Hepburn','Japanese','Literally',
                'Mandarin','McCune–Reischauer','Original title','Polish',
                'Revised Romanization','Romanized','Russian',
                'Simplified','Traditional','Yiddish']:
        if key in movie:
            alt_titles[key] = movie[key]
            movie.pop(key)


            return movie

#### Step 3: After looping through every key, add the alternative titles dict to the movie object.

In [44]:
def clean_movie(movie):
    movie = dict(movie) #create a non-destructive copy
    alt_titles = {}
    for key in ['Also known as','Arabic','Cantonese','Chinese','French',
                'Hangul','Hebrew','Hepburn','Japanese','Literally',
                'Mandarin','McCune–Reischauer','Original title','Polish',
                'Revised Romanization','Romanized','Russian',
                'Simplified','Traditional','Yiddish']:
        if key in movie:
            alt_titles[key] = movie[key]
            movie.pop(key)
    if len(alt_titles) > 0:
        movie['alt_titles'] = alt_titles

    return movie

We can make a list of cleaned movies with a list comprehension:

In [45]:
clean_movies = [clean_movie(movie) for movie in wiki_movies]

Set wiki_movies_df to be the DataFrame created from clean_movies, and print out a list of the columns.

In [46]:
wiki_movies_df = pd.DataFrame(clean_movies)
sorted(wiki_movies_df.columns.tolist())

['Adaptation by',
 'Animation by',
 'Audio format',
 'Based on',
 'Box office',
 'Budget',
 'Cinematography',
 'Color process',
 'Composer(s)',
 'Country',
 'Country of origin',
 'Created by',
 'Directed by',
 'Director',
 'Distributed by',
 'Distributor',
 'Edited by',
 'Editor(s)',
 'Executive producer(s)',
 'Followed by',
 'Genre',
 'Label',
 'Language',
 'Length',
 'Music by',
 'Narrated by',
 'Original language(s)',
 'Original network',
 'Original release',
 'Picture format',
 'Preceded by',
 'Produced by',
 'Producer',
 'Producer(s)',
 'Production company(s)',
 'Production location(s)',
 'Productioncompanies ',
 'Productioncompany ',
 'Recorded',
 'Release date',
 'Released',
 'Running time',
 'Screen story by',
 'Screenplay by',
 'Starring',
 'Story by',
 'Suggested by',
 'Theme music composer',
 'Venue',
 'Voices of',
 'Written by',
 'alt_titles',
 'imdb_link',
 'title',
 'url',
 'year']